### PR TITLE
Redux Container Pattern and Saga 

### PR DESCRIPTION
**Pull request by HassanSE19**

### CODE REVIEW

File: src/containers/TaskContainer/index.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  setTaskStatus: (targetIndex: number) => {
+    dispatch({
+      type: TODO_ACTIONS_TYPES.SET_TASK_STATUS,
+      payload: { targetIndex },
+    });
+  },
+});
+
+const AddTaskContainer = connect(mapStateToProps, mapDispatchToProps)(Task);
+export default AddTaskContainer;
```
Comment (mtalha-mergestack) [inline src/containers/TaskContainer/index.ts]: Isn't this Task container

File: src/store/saga.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import { all } from "redux-saga/effects";
+import { watchTodoListSaga } from "./taskList/taskListSaga";
+
+function* rootSaga() {
+  yield all([watchTodoListSaga()]);
+}
+
+export default rootSaga;
```
Comment (mtalha-mergestack) [inline src/store/saga.ts:8]: Add end of line.

File: src/store/taskList/taskListSaga.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  DELETE_TASK_COMPLETED,
+  DELETE_TASK_REJECTED,
+  SET_TASK_STATUS_STARTED,
+  SET_TASK_STATUS_COMPLETED,
+  SET_TASK_STATUS_REJECTED,
+} from "./taskListSlice";
+
+function* addTask({ payload }: PayloadAction<ITaskObj>) {
+  try {
+    yield put(ADD_TASK_STARTED());
```
Comment (mtalha-mergestack) [inline src/store/taskList/taskListSaga.ts]: We don't need to add started here

File: src/store/taskList/taskListSaga.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  } catch (error: any) {
+    yield put(ADD_TASK_REJECTED(error));
+  }
+}
+
+function* editTask({
+  payload: { newDesc, targetIndex },
+}: PayloadAction<IEditTaskPayload>) {
+  try {
+    yield put(EDIT_TASK_STARTED());
```
Comment (mtalha-mergestack) [inline src/store/taskList/taskListSaga.ts]: same as above

File: src/store/taskList/taskListSaga.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  } catch (error: any) {
+    yield put(EDIT_TASK_REJECTED(error));
+  }
+}
+
+function* deleteTask({
+  payload: { targetIndex },
+}: PayloadAction<IDeleteAndTogglePayload>) {
+  try {
+    yield put(DELETE_TASK_STARTED());
```
Comment (mtalha-mergestack) [inline src/store/taskList/taskListSaga.ts]: same

File: src/store/taskList/taskListSaga.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+  } catch (error: any) {
+    yield put(DELETE_TASK_REJECTED(error));
+  }
+}
+
+function* setTaskStatus({
+  payload: { targetIndex },
+}: PayloadAction<IDeleteAndTogglePayload>) {
+  try {
+    yield put(SET_TASK_STATUS_STARTED());
```
Comment (mtalha-mergestack) [inline src/store/taskList/taskListSaga.ts]: same

File: src/containers/TaskContainer/index.ts (hunk 1 above)
Comment (HassanSE19) [reply src/containers/TaskContainer/index.ts]: Nope, this container is only responsible for adding task and is built around add task component. 
Task component is separate and has modular functions like "EDIT", "DELETE" and "TOGGLE" status. And hence has its own container

File: src/utils/constants/actionTypes.ts [hunk 7]
```diff
@@ -0,0 +1,6 @@
+export enum TODO_ACTIONS_TYPES {
+  ADD_TASK = "ADD_TASK",
+  DELETE_TASK = "DELETE_TASK",
+  EDIT_TASK = "EDIT_TASK",
+  SET_TASK_STATUS = "SET_TASK_STATUS",
+}
```
Comment (mtalha-mergestack) [inline src/utils/constants/actionTypes.ts]: No need for these types

File: src/store/taskList/taskListSaga.ts [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+  } catch (error: any) {
+    yield put(SET_TASK_STATUS_REJECTED(error));
+  }
+}
+
+export function* watchTodoListSaga() {
+  yield takeEvery(TODO_ACTIONS_TYPES.ADD_TASK, addTask);
+  yield takeEvery(TODO_ACTIONS_TYPES.EDIT_TASK, editTask);
+  yield takeEvery(TODO_ACTIONS_TYPES.DELETE_TASK, deleteTask);
+  yield takeEvery(TODO_ACTIONS_TYPES.SET_TASK_STATUS, setTaskStatus);
```
Comment (mtalha-mergestack) [inline src/store/taskList/taskListSaga.ts:79]: instead of Types pass the Started ACTION types

File: src/containers/AddTaskContainer/index.ts [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import { connect } from "react-redux";
+import { AppDispatch } from "store/store";
+import AddTask from "components/AddTask";
+import { TODO_ACTIONS_TYPES } from "utils/constants/actionTypes";
+import { ITaskObj } from "type";
+
+const mapDispatchToProps = (dispatch: AppDispatch) => ({
+  addTask: (newTask: ITaskObj) => {
+    dispatch({ type: TODO_ACTIONS_TYPES.ADD_TASK, payload: newTask });
```
Comment (mtalha-mergestack) [inline src/containers/AddTaskContainer/index.ts:9]: PASS STARTED ACTION as type

File: src/containers/TaskContainer/index.ts [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+import { TODO_ACTIONS_TYPES } from "utils/constants/actionTypes";
+
+const mapStateToProps = (state: RootState, ownProps: ITaskContainerProps) => ({
+  task: ownProps.task,
+  index: ownProps.index,
+});
+
+const mapDispatchToProps = (dispatch: AppDispatch) => ({
+  editTask: (editData: IEditActionType) => {
+    dispatch({ type: TODO_ACTIONS_TYPES.EDIT_TASK, payload: editData });
```
Comment (mtalha-mergestack) [inline src/containers/TaskContainer/index.ts:14]: same started action

File: src/containers/TaskContainer/index.ts [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+  },
+  deleteTask: (targetIndex: number) => {
+    dispatch({
+      type: TODO_ACTIONS_TYPES.DELETE_TASK,
```
Comment (mtalha-mergestack) [inline src/containers/TaskContainer/index.ts:18]: statrted action

File: src/containers/TaskContainer/index.ts [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+      payload: { targetIndex },
+    });
+  },
+  setTaskStatus: (targetIndex: number) => {
+    dispatch({
+      type: TODO_ACTIONS_TYPES.SET_TASK_STATUS,
```
Comment (mtalha-mergestack) [inline src/containers/TaskContainer/index.ts:24]: same

File: src/store/taskList/taskListSaga.ts (hunk 3 above)
Comment (HassanSE19) [reply src/store/taskList/taskListSaga.ts]: So then how would we be setting loading state to true before we do some task operation.
We prepare our ready to set new array inside generator function after setting loading state true. We then distpatch the completed action if our operation was a success... 
One way would be to make a generic action that "START_ACTION" that would set isLoading true

**Pull request merged**